File: Xenon/Xenon/OptionsMenuState.ts
import CGameState = require("GameState");
import gsCMenu = require("Menu");
import gsCPoint = require("Point");
import Options = require("Options")
import enums = require("Enums");
import CControlMenuState = require("ControlMenuState");
import CVideoMenuState = require("VideoMenuState");
import CAudioMenuState = require("AudioMenuState");
import CMainMenuState = require("MainMenuState");

class COptionsMenuState extends CGameState {

    m_instance = null;
    m_menu: gsCMenu;
    m_options: Options;
    m_controlMenuState: CControlMenuState;
    m_videoMenuState: CVideoMenuState;
    m_audioMenuState: CAudioMenuState;
    m_mainMenuState: CMainMenuState;

    constructor() {
        super();
    }

    //-------------------------------------------------------------

    public instance(): CGameState {
        if (!this.m_instance)
            this.m_instance = new COptionsMenuState();
        return this.m_instance;
    }

    //-------------------------------------------------------------

    public create(): boolean {

        this.m_menu.clear();
        this.m_menu.addSelection("Control Options");
        this.m_menu.addSelection("Video Options");
        this.m_menu.addSelection("Audio Options");

        //this.m_menu.addSeperator();
        this.m_menu.addSelection("Back To Main Menu");
        this.m_menu.setWrap(true);
        this.m_menu.setPosition(new gsCPoint(0, 150));
        this.m_menu.setSpacing(new gsCPoint(0, 30));
        //this.m_menu.setCurrentItem(OM_BACK);
        this.m_menu.setFont(this.m_medium_font);

        return true;
    }

    //-------------------------------------------------------------

    public update(): boolean {

        //if (!CGameState::update())
        //return false;

        if (this.m_options.getOption(enums.OptionType.OPTION_BACKDROP)) {
            //m_backdrop.draw(new gsCPoint(0, 0)); 
        }
        else {
            //m_screen.clear(gsCColour(gsBLACK));
        }

        //m_starfield.move(4);
        //m_starfield.draw();

        this.m_medium_font.setTextCursor(new gsCPoint(0, 50));
        this.m_medium_font.justifyString("Options Menu");

        //this.m_menu.draw(ctx);
        //this.m_screen.flip();

        var item: enums.OptionsMenuItem = <enums.OptionsMenuItem>this.m_menu.getCurrentItem();

        //switch (getKey()) {
        //case gsKEY_RETURN:
        //case gsKEY_ENTER:
        //case gsKEY_LCONTROL:
        switch (item) {
            case enums.OptionsMenuItem.OM_CONTROL:
                //CGameState::playSample(SAMPLE_MENU_SELECTION);
                return this.changeState(this.m_controlMenuState.instance());
            case enums.OptionsMenuItem.OM_VIDEO:
                //CGameState::playSample(SAMPLE_MENU_SELECTION);
                return this.changeState(this.m_videoMenuState.instance());
            case enums.OptionsMenuItem.OM_AUDIO:
                //CGameState::playSample(SAMPLE_MENU_SELECTION);
                return this.changeState(this.m_audioMenuState.instance());
            case enums.OptionsMenuItem.OM_BACK:
                //CGameState::playSample(SAMPLE_MENU_BACK);
                return this.changeState(this.m_mainMenuState.instance());
        }
        //break;
        //case gsKEY_UP:
        //CGameState::playSample(SAMPLE_MENU_SELECTION);
        //this.m_menu.scroll(-1);
        //break;
        //        case gsKEY_DOWN:
        //CGameState::playSample(SAMPLE_MENU_SELECTION);
        //this.m_menu.scroll(1);
        //break;
        //}

        return true;
    }

    //-------------------------------------------------------------
}
export = COptionsMenuState;

File: Xenon-Original_C++_Code/xenon/source/optionsmenustate.cpp
//-------------------------------------------------------------
//
// Class:	COptionsMenuState
//
// Author:	John M Phillips
//
// Started:	06/05/00
//
// Base:	CGameState
//
// Derived:	None
//
//-------------------------------------------------------------

#include "game.h"

//-------------------------------------------------------------

COptionsMenuState *COptionsMenuState::m_instance = 0;

//-------------------------------------------------------------

COptionsMenuState::COptionsMenuState()
{
}

//-------------------------------------------------------------

COptionsMenuState::~COptionsMenuState()
{
}

//-------------------------------------------------------------

CGameState *COptionsMenuState::instance()
{
	if (!m_instance)
		m_instance = new COptionsMenuState;

	return m_instance;
}

//-------------------------------------------------------------

bool COptionsMenuState::create()
{
	m_menu.clear();

	m_menu.addSelection("Control Options");
	m_menu.addSelection("Video Options");
	m_menu.addSelection("Audio Options");

	m_menu.addSeperator();
	m_menu.addSelection("Back To Main Menu");

	m_menu.setWrap(true);
	m_menu.setPosition(gsCPoint(0,150));
	m_menu.setSpacing(gsCPoint(0,30));
	m_menu.setCurrentItem(OM_BACK);
	m_menu.setFont(&m_medium_font);

	return true;
}

//-------------------------------------------------------------

bool COptionsMenuState::update()
{
	if (!CGameState::update())
		return false;

	if (Options.getOption(OPTION_BACKDROP))
		m_backdrop.draw(gsCPoint(0,0));
	else
		m_screen.clear(gsCColour(gsBLACK));

	m_starfield.move(4);
	m_starfield.draw();

	m_medium_font.setTextCursor(gsCPoint(0,50));
	m_medium_font.justifyString("Options Menu");

	m_menu.draw();
	
	m_screen.flip();

	OptionsMenuItem item = (OptionsMenuItem) m_menu.getCurrentItem();

	switch (getKey()) {
		case gsKEY_RETURN:
		case gsKEY_ENTER:
		case gsKEY_LCONTROL:
			switch (item) {
				case OM_CONTROL:
					CGameState::playSample(SAMPLE_MENU_SELECTION);
					return changeState(CControlMenuState::instance());
				case OM_VIDEO:
					CGameState::playSample(SAMPLE_MENU_SELECTION);
					return changeState(CVideoMenuState::instance());
				case OM_AUDIO:
					CGameState::playSample(SAMPLE_MENU_SELECTION);
					return changeState(CAudioMenuState::instance());
				case OM_BACK:
					CGameState::playSample(SAMPLE_MENU_BACK);
					return changeState(CMainMenuState::instance());
				}
			break;
		case gsKEY_UP:
			CGameState::playSample(SAMPLE_MENU_SELECTION);
			m_menu.scroll(-1);
			break;
		case gsKEY_DOWN:
			CGameState::playSample(SAMPLE_MENU_SELECTION);
			m_menu.scroll(1);
			break;
		}

	return true;
}

//-------------------------------------------------------------

